### PR TITLE
Brimdemons

### PR DESCRIPTION
**Pull request by Fikou**
## About The Pull Request

Adds brimdemons, these are lavaland creatures firing blood lasers at their enemies.
Brimdemons drop brimdust, which can be used as a botany reagent (it heals the plant, removes weeds and pests and increases potency), and also to craft ore sensors, which are basically primitive ore scanners, on your ears.
Adds tropical lobstrosities to lavaland and removes them from xenobiology, also gives them a crusher trophy that staggers enemies for 3 seconds.
example image
image

## Why It's Good For The Game

More variety on lavaland is nice.

## Changelog

:cl: Fikou, Armhulen, smartkar
add: Brimdemons have invaded Lavaland.
del: Lobstrocities have awakened from their lavaland slumber, this somehow made them unavailable in the gold slime pool.
/:cl:

### CODE REVIEW

**Comment by bluetshirtguy** [comment]
That's some awesome stuff! Thanks fikou!

**Comment by D4C-420** [comment]
every miner upon first encountering this:
https://www.youtube.com/watch?v=HMR0Aldn6WU

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+	return ..()
+
+/mob/living/simple_animal/hostile/asteroid/brimdemon/Move(atom/newloc, dir, step_x , step_y)
+	if(firing)
+		return FALSE
+	return ..()
+
+/mob/living/simple_animal/hostile/asteroid/brimdemon/OpenFire()
+	if(firing)
+		to_chat(src, "<span class='warning'>You are already firing!</span>")
```
Comment (tf-4) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:83]:
```suggestion
		to_chat(src, span_warning("You are already firing!"))
```
same for the other spans

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm [hunk 2]
```diff
@@ -31,7 +31,7 @@
 	butcher_results = list(/obj/item/food/meat/crab = 2, /obj/item/stack/sheet/bone = 2)
 	robust_searching = TRUE
 	footstep_type = FOOTSTEP_MOB_CLAW
-	gold_core_spawnable = HOSTILE_SPAWN
+	crusher_loot = /obj/item/crusher_trophy/bear_paw
```
Comment (tf-4) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/lobstrosity.dm:34]:
```suggestion
	crusher_loot = /obj/item/crusher_trophy/lobster_claw
```

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm (hunk 1 above)
Comment (Fikou) [reply code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:83]: its april 2021 code ill go ahead and make it balloon alerts

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	icon_state = "brimdemon_fang"
+	desc = "A fang from a brimdemon's corpse."
+	denied_type = /obj/item/crusher_trophy/brimdemon_fang
+	var/static/list/comic_phrases = list("BOOM", "BANG", "KABLOW", "KAPOW", "OUCH", "BAM", "KAPOW", "WHAM", "POW", "KABOOM")
+
+/obj/item/crusher_trophy/brimdemon_fang/effect_desc()
+	return "mark detonation creates effects on the target"
+
+/obj/item/crusher_trophy/brimdemon_fang/on_mark_detonation(mob/living/target, mob/living/user)
+	target.balloon_alert_to_viewers("[pick(comic_phrases)]!")
```
Comment (JohnFulpWillard) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:188]: Why not just say?

Comment (Fikou) [reply code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:188]: because they arent saying it its a visual effect

**Comment by VladinXXV** [comment]
Why remove the lobstrosity from xenobio? The complaints today were from one person, based on them thinking that you couldn't kill them by normal means, which I then proceeded to prove false, killing both lobstrosity variants with nothing other than a toolbox (later repeated with a circular saw).

**Comment by Fikou** [comment]
> Why remove the lobstrosity from xenobio? The complaints today were from one person, based on them thinking that you couldn't kill them by normal means, which I then proceeded to prove false, killing both lobstrosity variants with nothing other than a toolbox (later repeated with a circular saw).

legions and goliaths dont and i think we shouldnt have as many lavaland fauna (that can break walls) in the pool

**Comment by VladinXXV** [comment]
> > Why remove the lobstrosity from xenobio? The complaints today were from one person, based on them thinking that you couldn't kill them by normal means, which I then proceeded to prove false, killing both lobstrosity variants with nothing other than a toolbox (later repeated with a circular saw).
> 
> legions and goliaths dont and i think we shouldnt have as many lavaland fauna (that can break walls) in the pool

Fair, I reckon if you really need a mob that breaks walls you can upgrade the mob of choice by rolling a fugu.

**Comment by HugoOdaX** [comment]
someone has been playing a lot of TBOI recently

**Comment by CRITAWAKETS** [comment]
that laser beam and death sprite look so fucking jank i love em

honestly, dont resprite, they're perfect

**Comment by cacogen** [comment]
You should add removing the lobstrosity from the gold slime core spawn pool to the changelog

**Comment by Shadowflame909** [comment]
Who would win in a fight of simple mob beams, the brimdemons or the seedlings?

Also, I'd like to mention that tropical lobstrocities already exist as a mining fauna on icebox. That change isn't that big of a deal. It's just already in place on certain maps.

**Comment by Fikou** [comment]
> Who would win in a fight of simple mob beams, the brimdemons or the seedlings?
> 
> Also, I'd like to mention that tropical lobstrocities already exist as a mining fauna on icebox. That change isn't that big of a deal. It's just already in place on certain maps.

no thats ice lobstrosities not tropical ones

**Review by Imaginos16 — APPROVED** [review]
its ok spritewise

File: code/datums/components/crafting/recipes.dm [hunk 4]
```diff
@@ -948,6 +948,15 @@
 	result = /obj/item/reagent_containers/glass/bucket/wooden
 	category = CAT_PRIMAL
 
+/datum/crafting_recipe/ore_sensor
+	name = "Ore Sensor"
+	time = 3 SECONDS
+	reqs = list(/datum/reagent/brimdust = 15,
+		/obj/item/stack/sheet/bone = 1,
+		/obj/item/stack/sheet/sinew = 1)
```
Comment (tralezab) [inline code/datums/components/crafting/recipes.dm:956]: ending paren to the next line, unindented

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+			if(potential_block.opacity)
+				blocked = TRUE
+				break
+		if(blocked)
+			break
+		var/atom/new_brimbeam = new /obj/effect/brimbeam(affected_turf)
+		new_brimbeam.dir = dir
+		beamparts += new_brimbeam
+		for(var/mob/living/hit_mob in affected_turf.contents)
+			if(istype(hit_mob, /mob/living/simple_animal/hostile/asteroid/brimdemon))
```
Comment (tralezab) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:132]: faction check probably makes more sense in this case, no?

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+
+/obj/item/crusher_trophy/brimdemon_fang
+	name = "brimdemon's fang"
+	icon_state = "brimdemon_fang"
+	desc = "A fang from a brimdemon's corpse."
+	denied_type = /obj/item/crusher_trophy/brimdemon_fang
+	var/static/list/comic_phrases = list("BOOM", "BANG", "KABLOW", "KAPOW", "OUCH", "BAM", "KAPOW", "WHAM", "POW", "KABOOM")
+
+/obj/item/crusher_trophy/brimdemon_fang/effect_desc()
+	return "mark detonation creates effects on the target"
```
Comment (tralezab) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:191]: a little more descript than this, surely

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+			hit_mob.adjustFireLoss(25)
+			to_chat(hit_mob, span_userdanger("You're hit by [src]'s brimbeam!"))
+	if(length(beamparts))
+		var/atom/last_brimbeam = beamparts[length(beamparts)]
+		last_brimbeam.icon_state = "brimbeam_end"
+		var/atom/first_brimbeam = beamparts[1]
+		first_brimbeam.icon_state = "brimbeam_start"
+	addtimer(CALLBACK(src, .proc/end_laser), 2 SECONDS)
+
+/mob/living/simple_animal/hostile/asteroid/brimdemon/proc/end_laser()
```
Comment (tralezab) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:143]: doc it!

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+	. = ..()
+	check_fire()
+
+/mob/living/simple_animal/hostile/asteroid/brimdemon/proc/check_fire()
+	if(key || QDELETED(target) || get_dist(src, target) > BRIMBEAM_RANGE || !(get_dir(src, target) in GLOB.cardinals))
+		return
+	face_atom(target)
+	OpenFire()
+
+/mob/living/simple_animal/hostile/asteroid/brimdemon/proc/fire_laser()
```
Comment (tralezab) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:108]: doc it!

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+	visible_message(span_danger("[src] starts charging!"))
+	balloon_alert(src, "charging...")
+	addtimer(CALLBACK(src, .proc/fire_laser), 1 SECONDS)
+	COOLDOWN_START(src, ranged_cooldown, ranged_cooldown_time)
+
+/mob/living/simple_animal/hostile/asteroid/brimdemon/Moved()
+	. = ..()
+	check_fire()
+
+/mob/living/simple_animal/hostile/asteroid/brimdemon/proc/check_fire()
```
Comment (tralezab) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:102]: doc it!

File: code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+/obj/effect/brimbeam/Destroy()
+	STOP_PROCESSING(SSfastprocess, src)
+	return ..()
+
+/obj/effect/brimbeam/process()
+	for(var/mob/living/hit_mob in get_turf(src))
+		damage(hit_mob)
+
+/obj/effect/brimbeam/proc/damage(mob/living/hit_mob)
+	if(istype(hit_mob, /mob/living/simple_animal/hostile/asteroid/brimdemon))
```
Comment (tralezab) [inline code/modules/mob/living/simple_animal/hostile/mining_mobs/brimdemon.dm:178]: faction check

**Review by tralezab — CHANGES_REQUESTED** [review]

**Pull request merged**